### ISSUE TITLE
Can I split the file but keep heading in each file?

### ISSUE THREAD

**Issue by xiaodaigh** [privefl/bigreadr#2]
Currently, the heading is dropped in all but the first file. It would be good to have an option to keep heading in all the files.

**Comment by privefl** [comment]
Can you please be more precise?
E.g. give some code example.

**Comment by xiaodaigh** [comment]
```r
write.csv(cars, "cars.csv")

a = bigreadr::split_file("cars.csv", every_nlines = 10)
f = bigreadr::get_split_files(a)

read.csv(f[1])
# X speed dist
# 1 1     4    2
# 2 2     4   10
# 3 3     7    4
# 4 4     7   22
# 5 5     8   16
# 6 6     9   10
# 7 7    10   18
# 8 8    10   26
# 9 9    10   34


read.csv(f[2])
# X10 X11 X17
# 1  11  11  28
# 2  12  12  14
# 3  13  12  20
# 4  14  12  24
# 5  15  12  28
# 6  16  13  26
# 7  17  13  34
# 8  18  13  34
# 9  19  13  46
```

Can we have it so that f[2] also has the header speed and dist written to the file when splitting?

**Comment by privefl** [comment]
Hum, I see.
The idea here is that the file is just splitted.
So, if there is some header, then it would appear in the first split only.

I don't think it would be too difficult to implement; do you want to try to do it with a PR?
Otherwise, I think I would have one hour this WE for this.

What is your use-case?

**Comment by xiaodaigh** [comment]
I need to pass in colClasses and col.names if the file doesn't have header.
And in fread you can't use both at the same time.

**Comment by privefl** [comment]
You want to get the header written in all files? 
Or just to read all parts in R (as e.g. data frames) with a header (colnames)?

**Comment by xiaodaigh** [comment]
Write header to files. That's the most flexible.

**Comment by privefl** [comment]
Not sure it is the most flexible. Nor the simplest (would need to change C++ code instead of R code).

See how I read colnames for the other parts: 
https://github.com/privefl/bigreadr/blob/master/R/read.R#L143-L154

**Comment by xiaodaigh** [comment]
Also now try to read `col.names` AND `colClasses` at the same time, for a headerless file (which are the parts after the first).

It's not possible to do it by specifying `colClasses` using  `col.names` as strings https://github.com/Rdatatable/data.table/issues/3816 if the file is headerless.

You have to use [some tricks](https://stackoverflow.com/questions/57782386/how-do-i-use-col-names-and-colclasses-together-in-data-tablefread/57784101#57784101) to do it which are super user unfriendly. 

The below MWE gives error.
```r
data.table::fwrite(cars, "cars.csv")
a = bigreadr::split_file("cars.csv", every_nlines = 10)
f = bigreadr::get_split_files(a)

data.table::fread(
  f[1],
  colClasses = c(speed="character", dist = "character"))


data.table::fread(
  f[2], 
  col.names = c("speed","dist"),
  colClasses = c(speed="character", dist = "character"))
```
and the error is 

```r
Error in data.table::fread(f[2], col.names = c("speed", "dist"), colClasses = c(speed = "character",  : 
  Column name 'speed' in colClasses[[1]] not found
```

Basically, you can't use `col.names` and `colClasses` together at the same time. Again, very bad user experience.

**Comment by privefl** [comment]
I think, you can just use `data.table::fread(f[2], colClasses = c("character", "character"))`.

**Comment by xiaodaigh** [comment]
Your example is fine for two columns file. Imagine reading a file with
thousands of columns. And I only know the names of the columns but not the
position. Nwo I have to find the position before using it. Pretty painful.

**Comment by privefl** [comment]
With e.g. a `match()` on the names.
If you share a subset of your file, I can help you with this.

I see that this can be useful. I'll take some time this WE to look if I can easily change the C++ code to provide this option.

**Comment by xiaodaigh** [comment]
I made a pr see #3

**Comment by xiaodaigh** [comment]
Thanks for submitting it to CRAN. disk.frame now imports on bigreadr 
![image](https://user-images.githubusercontent.com/4497189/65092630-278e4080-da00-11e9-8ce2-a415da660fae.png)

**Comment by privefl** [comment]
Nice to hear.
If you need anything else, please let me know.
Yet, I won't have time to help you with {disk.frame} directly. I'm sorry about that.

PS: I've just submitted a new version of {bigreadr} that allows automatic expanding of paths with e.g. `~/` at the start (this is not done automatically in C).

PS2: I was reading some of your vignettes and wondered if we could read and process files in parallel. Basically, split with a master process, and read & process with other cores as soon as one block is ready.
